kamon-core/src/main/scala/kamon/context/Storage.scala

```diff
@@ -27,28 +29,45 @@ object Storage {
     def close(): Unit
   }
 
-
+  /**
+    * Wrapper that implements optimized {@link ThreadLocal} access pattern ideal for heavily used
+    * ThreadLocals.
+    *
+    * It is faster to use a mutable holder object and always perform ThreadLocal.get() and never use
+    * ThreadLocal.set(), because the value is more likely to be found in the ThreadLocalMap direct hash
+    * slot and avoid the slow path ThreadLocalMap.getEntryAfterMiss().
+    *
+    * Important: this thread local will live in ThreadLocalMap forever, so use with care.
+    */
   class ThreadLocal extends Storage {
-    private val tls = new java.lang.ThreadLocal[Context]() {
-      override def initialValue(): Context = Context.Empty
+    private val tls = new java.lang.ThreadLocal[ContextHolder]() {
+      override def initialValue() = new ContextHolder(Context.Empty)
```

Comment:
I like this but won't it risk a classloader leak? ContextHolder will prevent the classloader from being unloadable unless it is explicitly cleaned. maybe this doesn't mean anything if it is loaded via an agent or the classloader is never expected to be unloaded...

Reply:
@adriancole thanks for pointing that, yes you are right, here we can have a classloader leak if the context is not correctly closed, what do you think about something like this:     

``` scala
final class ContextHolder(var value:Context) extends WeakReference(value)

```

Reply:
@adriancole I've run the benchmarks with yours proposed changes:

![image](https://user-images.githubusercontent.com/2567525/46117147-0d979480-c1d6-11e8-9175-a40020d05937.png)

looks really good!!

Reply:
👍